### PR TITLE
Test [[CanBlock]] in various agents

### PR DESCRIPTION
**Pull request by domenic**
Follows https://github.com/whatwg/html/pull/2521. This downgrades the resources submodule due to https://github.com/w3c/web-platform-tests/issues/5568 which prevents these tests (and many, many other tests in WPT) from working.

FYI @binji @mtrofin on the Blink side, @lars-t-hansen on the Gecko side. Failures:

- Chrome allows atomics in service workers, against the emerging spec. This is not likely to change.
   - Filed https://bugs.chromium.org/p/chromium/issues/detail?id=711809
- Firefox throws the wrong error type in windows and service workers
  - Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1356697
- Firefox disallows atomics in shared workers, against the emerging spec. This is still under some minor discussion in the PR.
- Safari passes the applicable two tests (window and dedicated workers)

Worklets are not tested at this time due to spec uncertainty.

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/5569)

### CODE REVIEW

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5569/)

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/222219977)

# Lint #

Passed

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/222219979)

# Firefox (nightly channel) #
Testing web-platform-tests at revision c155768f1dd1ba179ab19e6d0cd97ac1cbe75b13
Using browser at version BuildID 20170414100243; SourceStamp cda24082bff8864a6e53726feeae33cae9e17309
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>4 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5569/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-dedicatedworker.html">/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-dedicatedworker.html</a></summary>


|                    Subtest                     | Results | Messages |
|------------------------------------------------|---------|----------|
|                                                | OK      |          |
| `[[CanBlock]] in a DedicatedWorkerGlobalScope` | PASS    | `{}`     |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5569/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-serviceworker.html">/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-serviceworker.html</a></summary>


|           Subtest           | Results |                                                     Messages                                                     |
|-----------------------------|---------|------------------------------------------------------------------------------------------------------------------|
|                             | OK      |                                                                                                                  |
| `Service worker test setup` | FAIL    | `assert_unreached: unregister and register should not fail: The operation is insecure. Reached unreachable code` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5569/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-sharedworker.html">/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-sharedworker.html</a></summary>


|                   Subtest                   | Results |                 Messages                |
|---------------------------------------------|---------|-----------------------------------------|
|                                             | OK      |                                         |
| `[[CanBlock]] in a SharedWorkerGlobalScope` | FAIL    | `waiting is not allowed on this thread` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5569/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-window.html">/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-window.html</a></summary>


|                     Subtest                     | Results |                                                                                                                           Messages                                                                                                                           |
|-------------------------------------------------|---------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                 | OK      |                                                                                                                                                                                                                                                              |
| `[[CanBlock]] in a similar-origin window agent` | FAIL    | `uncaught exception: Error: assert_throws: Atomics.wait must throw in a window context function "() => {\n  Atomics.wait(ta, 0, 0, 10);\n}" threw object "Error: waiting is not allowed on this thread" ("Error") expected object "TypeError" ("TypeError")` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/222219980)

# Chrome (unstable channel) #
Testing web-platform-tests at revision c155768f1dd1ba179ab19e6d0cd97ac1cbe75b13
Using browser at version 59.0.3067.0 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>4 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5569/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-dedicatedworker.html">/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-dedicatedworker.html</a></summary>


|                    Subtest                     | Results |              Messages              |
|------------------------------------------------|---------|------------------------------------|
|                                                | OK      |                                    |
| `[[CanBlock]] in a DedicatedWorkerGlobalScope` | FAIL    | `SharedArrayBuffer is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5569/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-serviceworker.html">/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-serviceworker.html</a></summary>


|           Subtest           | Results |                                                                Messages                                                                |
|-----------------------------|---------|----------------------------------------------------------------------------------------------------------------------------------------|
|                             | OK      |                                                                                                                                        |
| `Service worker test setup` | FAIL    | `assert_unreached: unregister should not fail: Only secure origins are allowed (see: https://goo.gl/Y0ZkNV). Reached unreachable code` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5569/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-sharedworker.html">/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-sharedworker.html</a></summary>


|                   Subtest                   | Results |              Messages              |
|---------------------------------------------|---------|------------------------------------|
|                                             | OK      |                                    |
| `[[CanBlock]] in a SharedWorkerGlobalScope` | FAIL    | `SharedArrayBuffer is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5569/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-window.html">/html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/canblock-window.html</a></summary>


|                     Subtest                     | Results |                           Messages                          |
|-------------------------------------------------|---------|-------------------------------------------------------------|
|                                                 | OK      |                                                             |
| `[[CanBlock]] in a similar-origin window agent` | FAIL    | `Uncaught ReferenceError: SharedArrayBuffer is not defined` |

</details>
</details>

File: html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/worker-that-requires-success.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+"use strict";
+importScripts("/resources/testharness.js");
+
+test(() => {
+  const sab = new SharedArrayBuffer(16);
+  const ta = new Int32Array(sab);
+
+  // Test passes if this doesn't throw
+  Atomics.wait(ta, 0, 0, 10);
```
Comment (annevk) [inline html/webappapis/scripting/processing-model-2/integration-with-the-javascript-agent-formalism/worker-that-requires-success.js]: Can we assert the return value to be "ok"?

**Review by annevk — COMMENTED** [review]
This looks okay in general, but I wish we could annotate these tests somehow for removal if we ever get proper test262 integration going.

Also, should the `.js` resources be in a resources directory? It's never clear to me what the convention is there, if anything.

**Comment by domenic** [comment]
@binji @lars-t-hansen can you help get this test passing? It seems that in a dedicated worker, in both Chrome and Firefox, the following code gives "timed-out" instead of "ok":

```
"use strict";
importScripts("/resources/testharness.js");

test(() => {
  const sab = new SharedArrayBuffer(16);
  const ta = new Int32Array(sab);

  assert_equals(Atomics.wait(ta, 0, 0, 10), "ok");
}, `[[CanBlock]] in a ${self.constructor.name}`);

done();
```

**Comment by syg** [comment]
@domenic `Atomics.wait` would return "ok" if it's explicitly woken by `Atomics.wake`. Is there something that's waking it?

**Comment by domenic** [comment]
Oh, I thought it would return "ok" if the value successfully became the target value (which it already is). So "timed-out" is correct? Any suggestions for a better test?

Is it a Chrome bug that it returns "ok" in shared workers?

**Comment by syg** [comment]
@domenic "timed-out" is fine here, I think, if the point is that it doesn't throw. BTW if the TA value isn't equal to the supplied value, you get "not-equal".

**Comment by binji** [comment]
Hm, I'm not sure why this test would return "ok" in any context, it should either timeout or throw, as syg mentioned.

**Comment by lars-t-hansen** [comment]
What @syg said.  Additionally, observe that [[CanBlock]] is tested early so (a) a timeout of 0 is fine here, and you should still get "timed-out"; or (b) leaving off the timeout while first setting ta[0]=1 so that you get "not-equal" is also a valid test of [[CanBlock]].

**Comment by domenic** [comment]
Thanks for the help all. I just updated the test to require "timed-out". Chrome seems to be passing now; I'm not sure what was going on there locally, but maybe I had an old version of the test open or something and was being caught out by shared workers being... shared... between tabs.

**Comment by annevk** [comment]
Since service workers require HTTPS I renamed the resource to include .https. I believe that's the correct thing to do.

**Comment by annevk** [comment]
I had to force push because I had to rebase to get past the failing Chrome bot that was disabled by a newer commit. I'll let someone else double check before landing too.

See https://github.com/whatwg/html/pull/2521#issuecomment-297322650 for the browser bug analysis. Led to one bug against Firefox.

**Pull request merged**